flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java

```diff
@@ -426,11 +431,7 @@ public T deserialize(DataInputView source) throws IOException {
 
         if ((flags & IS_SUBCLASS) != 0) {
             String subclassName = source.readUTF();
-            try {
-                actualSubclass = Class.forName(subclassName, true, cl);
-            } catch (ClassNotFoundException e) {
-                throw new RuntimeException("Cannot instantiate class.", e);
-            }
+            actualSubclass = getSubclassByName(subclassName);
```

Comment:
3 thoughts arounds this:
- could there by a possibility that this method is driven such that different class loaders would be used to load the subclass? If so maybe it is safer to not cache or consider caching per classloader.
- I guess for a long running job, with lots of pojos and subclasses , the cache just grows, could this become an issue?
- I think we should have a unit test for this cache.

Reply:
1. Classloader: The same context classloader (stored in variable `cl`) is used within the same serializer instance here, so this should not be an issue here.
2. It is possible, though not quite likely. Similar issues apply to the existing `subclassSerializerCache` as well, a potential improvement is to use some capacity-limited cache impl like the Gauva cache.
3. For UT, the correctness is already covered by the existing `PojoSubclassSerializerTest`; for performance, a new benchmark is more suited, and I plan to add it in [FLINK-37681](https://issues.apache.org/jira/browse/FLINK-37681)